base/setup/lib/utils/filesup.c

```diff
@@ -71,78 +68,84 @@ SetupCreateSingleDirectory(
                           NULL,
                           0);
     if (NT_SUCCESS(Status))
-    {
         NtClose(DirectoryHandle);
-    }
-
-    RtlFreeUnicodeString(&PathName);
 
     return Status;
 }
 
 NTSTATUS
 SetupCreateDirectory(
-    IN PCWSTR PathName)
+    _In_ PCWSTR DirectoryName)
 {
     NTSTATUS Status = STATUS_SUCCESS;
-    PWCHAR PathBuffer = NULL;
+    UNICODE_STRING PathNameU;
+    PWCHAR PathBuffer;
     PWCHAR Ptr, EndPtr;
-    ULONG BackslashCount;
-    ULONG Size;
-
-    Size = (wcslen(PathName) + 1) * sizeof(WCHAR);
-    PathBuffer = RtlAllocateHeap(RtlGetProcessHeap(), HEAP_ZERO_MEMORY, Size);
-    if (PathBuffer == NULL)
-        return STATUS_INSUFFICIENT_RESOURCES;
 
-    wcscpy(PathBuffer, PathName);
-    EndPtr = PathBuffer + wcslen(PathName);
-
-    Ptr = PathBuffer;
+    if (!RtlCreateUnicodeString(&PathNameU, DirectoryName))
+        return STATUS_NO_MEMORY;
+    PathBuffer = PathNameU.Buffer;
+    EndPtr = PathBuffer + (PathNameU.Length / sizeof(WCHAR));
 
-    /* Skip the '\Device\HarddiskX\PartitionY\ part */
-    BackslashCount = 0;
-    while (Ptr < EndPtr && BackslashCount < 4)
+    /* Find the deepest existing sub-directory: start from the
+     * end and go back verifying each sub-directory in turn */
+    for (Ptr = EndPtr; PathBuffer < Ptr;)
     {
-        if (*Ptr == L'\\')
-            BackslashCount++;
-
-        Ptr++;
-    }
+        WCHAR wch = UNICODE_NULL;
+        BOOLEAN bExists;
 
-    while (Ptr < EndPtr)
-    {
-        if (*Ptr == L'\\')
+        /* If we are on a separator, save and erase the next character */
+        if ((Ptr < EndPtr) && (*Ptr == OBJ_NAME_PATH_SEPARATOR))
         {
-            *Ptr = 0;
+            wch = *(Ptr+1);
+            *(Ptr+1) = UNICODE_NULL;
+        }
 
-            DPRINT("PathBuffer: %S\n", PathBuffer);
-            if (!DoesDirExist(NULL, PathBuffer))
-            {
-                DPRINT("Create: %S\n", PathBuffer);
-                Status = SetupCreateSingleDirectory(PathBuffer);
-                if (!NT_SUCCESS(Status))
-                    goto done;
-            }
+        DPRINT("PathBuffer: %S\n", PathBuffer);
+        bExists = DoesDirExist(NULL, PathBuffer);
 
-            *Ptr = L'\\';
-        }
+        if (wch) /* Restore the erased character */
+            *(Ptr+1) = wch;
+
+        /* Stop if we've found the deepest existing sub-directory */
+        if (bExists)
+            break;
 
-        Ptr++;
+        /* Skip back path separators if any */
+        while ((PathBuffer < Ptr) && (*Ptr == OBJ_NAME_PATH_SEPARATOR))
```

Comment:
Same here, it is a bit weird to sometimes have a check `Ptr < EndPtr` and sometimes have a check `PathBuffer < Ptr`
(Where `Ptr` is in different places)

Reply:
I understand that, but in my opinion it makes it harder to read when the order of arguments changes.